features/crc/mbed_crc.h

```diff
@@ -0,0 +1,223 @@
+            return CRC_7;
+        } else if (CRC_8BIT_CCITT == polynomial) {
+            return CRC_8;
+        } else if ((CRC_16BIT_CCITT == polynomial) ||
+                (CRC_16BIT_SD == polynomial) ||
+                (CRC_16BIT_IBM == polynomial)) {
+            return CRC_16;
+        } else if (CRC_32BIT == polynomial) {
+            return CRC_32;
+        }
+        return CRC_INVALID;
+    }
+
+    /** Compute CRC for the data input
+     *
+     *  @param [IN]  buffer  Data bytes
+     *  @param [IN]  size  Size of data
+     *  @param [OUT] crc  CRC
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, crc_size_t *crc) = 0;
+
+    /** Compute partial start, indicate start of partial computation
+     *
+     *  This API should be called before performing any partial computation
+     *  with compute_partial API.
+     *
+     *  @param [OUT] crc  Initial CRC value set by the API
+     *  @return  0  on success or a negative when not supported
+     *  @note: CRC is an out parameter and must be reused with compute_partial
+     *         and compute_partial_stop without any modifications in application.
+     */
+    virtual int32_t compute_partial_start(crc_size_t *crc) const
+    {
+        return -1;
+    }
+
+    /** Compute partial CRC for the data input.
+     *
+     *  CRC data if not available fully, CRC can be computed in parts with available data.
+     *  Previous CRC output should be passed as argument to the current compute_partial call.
+     *  @pre: Call \ref compute_partial_start to start the partial CRC calculation.
+     *  @post: Call \ref compute_partial_stop to get the final CRC value.
+     *
+     *  @param[IN]  buffer  Data bytes
+     *  @param[IN]  size  Size of data
+     *  @param[OUT] crc  CRC value (intermediate CRC )
+     *  @return  0  on success or a negative error code on failure
+     *  @note: CRC as output in compute_partial is not final CRC value, call @ref compute_partial_stop
+     *         to get final correct CRC value.
+     */
+    virtual int32_t compute_partial(void *buffer, crc_data_size_t size, crc_size_t *crc)
+    {
+        return -1;
+    }
+
+    /** Get the final CRC value of partial computation.
+     *
+     *  CRC value available in partial computation is not correct CRC, as some
+     *  algorithms require remainder to be reflected and final value to be XORed
+     *  This API is used to perform final computation to get correct CRC value.
+     *
+     *  @param[OUT] crc  CRC result
+     */
+    virtual int32_t compute_partial_stop(crc_size_t *crc)
+    {
+        return -1;
+    }
+
+    /** Get the current CRC data size
+     *
+     * @return  CRC data size in bytes
+     */
+    uint8_t get_data_size(void) const
+    {
+        crc_width_t width = this->get_width();
+        return (width <= 8 ? 1 : (width <= 16 ? 2 : 4));
+    }
+
+    /** Get the top bit of current CRC
```

Comment:
I don't get that, what's the use case?

Reply:
Top bit will be used in software computation of CRC calculation. While performing the division operation we do not shift one bit at a time, the divisor is aligned to the first 1 of dividend. Here we perform right shifts and align dividend first 1 at the top-bit to perform division.

Reply:
That seems like we are exposing internal functions as API. Also 'top bit' is msb? maybe we should stick to msb/lsb?

Reply:
Top bit is not necessary MSB, in case of 6-bit CRC it will be 6th bit on unsigned char and not MSB bit.

Reply:
> That seems like we are exposing internal functions as API.

Yes, neither did I wanted them to be APIs (were implemented as macros earlier). But was suggested to have them as functions. I have two options 
1. Protected (as used by all derived classes in software implementation)
2. Move to software implementation class (multiple copies)

Reply:
Protected sounds good.